analysis/GreenSKU-Framework/notebooks/carbon_savings.ipynb

```diff
@@ -0,0 +1,667 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Plot carbon savings relative to 0 for each config\n",
+    "grid_mix = 0.35\n",
+    "sns.set_style(\"whitegrid\")\n",
+    "fig, ax = plt.subplots(figsize=(6, 1.5))\n",
+    "config_labels = [\n",
+    "    \"GreenSKU-Bergamo\",\n",
+    "    \"GreenSKU-CXL\",\n",
+    "    \"GreenSKU-Full\",\n",
+    "]\n",
+    "\n",
+    "blue = \"#0072B2\"\n",
+    "green = \"#009E73\"\n",
+    "orange = \"#E69F00\"\n",
+    "yellow = \"#E69F00\"\n",
+    "red = \"#D55E00\"\n",
+    "black = \"#000000\"\n",
+    "light_blue = \"#56B4E9\"\n",
+    "pink = \"#CC79A7\"\n",
+    "colors = [\n",
+    "    green,\n",
+    "    green,\n",
+    "    green\n",
+    "]\n",
+    "\n",
+    "linestyles = [\n",
+    "    \"solid\",\n",
+    "    \"dotted\",\n",
+    "    \"dashed\"\n",
+    "]\n",
+    "azure_dcs = [\n",
+    "    'ussouth',\n",
+    "    'useast',\n",
+    "]\n",
+    "google_dcs = [\n",
+    "    'us-south1',\n",
+    "    'us-east1'\n",
+    "]\n",
+    "azure_intensities = [get_dc_region_intensity(dc, azure_data) for dc in azure_dcs]\n",
+    "google_intensities = [get_dc_region_intensity(dc, google_data) for dc in google_dcs]\n",
+    "intensities = azure_intensities + google_intensities\n",
+    "intensity_colors = [\n",
+    "    \"purple\",\n",
+    "    \"darkorange\",\n",
+    "    \"blue\",\n",
+    "    \"black\",\n",
+    "]\n",
+    "annotations = [\n",
+    "    \"Azure\\nus-south\",\n",
+    "    \"Azure\\nus-east\",\n",
+    "    \"Google\\nus-south1\",\n",
+    "    \"Google\\nus-east1\"\n",
+    "]\n",
+    "\n",
+    "# Plot carbon savings for each config across carbon intensities\n",
+    "for i, config in enumerate(trace_config_data):\n",
+    "    ax.plot(CI_steps, \n",
+    "            config[\"carbon_savings\"], \n",
+    "            label=config_labels[i], \n",
+    "            color=colors[i], \n",
+    "            linestyle=linestyles[i], \n",
+    "            linewidth=2.2)\n",
+    "\n",
+    "# Plot carbon intensity annotations\n",
+    "for i, intensity in enumerate(intensities):\n",
+    "    ax.axvline(x=intensity, color=intensity_colors[i], linestyle=\"-\", alpha=0.2, linewidth=8)\n",
+    "    ax.annotate(annotations[i], xy=(intensity, 0.5), xycoords=\"data\", ha=\"center\", fontsize=9, color=intensity_colors[i], alpha=1)\n",
+    "\n",
+    "# Set plot parameters\n",
+    "ax.set_xlim([0, 0.4])\n",
+    "ax.set_ylim([0, 25])\n",
+    "ax.set_ylim(bottom=0)\n",
+    "# ax.set_yticks(np.arange(0, 30, 10))\n",
+    "ax.set_xlabel(\"Carbon intensity (kgCO$\\mathregular{_2}$e/kWh)\", fontsize=10)\n",
+    "ax.set_ylabel(\"Cluster carbon\\nsavings (%)\", fontsize=10)\n",
+    "\n",
+    "# Set legend above plot\n",
+    "ax.legend(loc=\"upper center\", bbox_to_anchor=(0.665, 1.02), ncol=2, fontsize=9)\n",
+    "\n",
+    "# Save plot to file\n",
+    "plt.savefig(\"../figures/generated_figures/Figure_12.png\", bbox_inches=\"tight\", dpi=500)"
```

Comment:
In addition to saving it, it would be nice to output all these figures so it shows in the notebook.